### PR TITLE
Should cache destination responses

### PR DESCRIPTION
**Pull request by miherlosev**
Cache resources (at this stage only scripts and styles) between sessions as it does browser: analyze caching headers and decide about sending a real HTTP request. The 'http-cache-semantics' module is used to make a decision about the necessity of refreshing request data.

### CODE REVIEW

**Comment by miherlosev** [comment]
FPR

File: src/request-pipeline/context.ts [hunk 1]
```diff
@@ -296,6 +299,13 @@ export default class RequestPipelineContext {
             .map(userScript => userScript.url);
     }
 
+    private async _getDestResBody (res: IncomingMessageLike | http.IncomingMessage | FileStream): Promise<Buffer> {
+        if (IncomingMessageLike.isIncomingMessageLike(res))
+            return (res as IncomingMessageLike).getBody();
+
+        return await fetchBody(this.destRes);
```
Comment (wentwrong) [inline src/request-pipeline/context.ts:306]: Minor: I think we can get rid of using `await` here. Reason: [Why Using `return await` Is a Bad Idea?](http://hassansin.github.io/Why-return-await-Is-a-Bad-Idea)

Comment (miherlosev) [reply src/request-pipeline/context.ts:306]: Ok

**Comment by miherlosev** [comment]
FPR

File: src/request-pipeline/incoming-message-like.ts [hunk 2]
```diff
@@ -52,4 +52,26 @@ export default class IncomingMessageLike extends Readable {
...
+        const { headers, trailers, statusCode } = res;
+
+        return new IncomingMessageLike({
+            headers,
+            trailers,
+            statusCode
+        });
+    }
+
+    static isIncomingMessageLike (obj: object): boolean {
```
Comment (wentwrong) [inline src/request-pipeline/incoming-message-like.ts:74]:
```suggestion
    static isIncomingMessageLike (obj: object): obj is IncomingMessageLike {
```

File: src/request-pipeline/context.ts [hunk 3]
```diff
@@ -296,6 +299,13 @@ export default class RequestPipelineContext {
             .map(userScript => userScript.url);
     }
 
+    private async _getDestResBody (res: IncomingMessageLike | http.IncomingMessage | FileStream): Promise<Buffer> {
+        if (IncomingMessageLike.isIncomingMessageLike(res))
+            return (res as IncomingMessageLike).getBody();
```
Comment (wentwrong) [inline src/request-pipeline/context.ts:304]:
```suggestion
            return res.getBody();
```
(related to the suggestion below)

**Review by wentwrong — APPROVED** [review]

File: src/request-pipeline/destination-request/index.ts [hunk 4]
```diff
@@ -58,31 +88,23 @@ export default class DestinationRequest extends EventEmitter implements Destinat
...
-                });
-            }
+            if (this.cache) {
+                const cachedResponse = requestCache.getResponse(this.opts);
+
+                if (cachedResponse) {
+                    // NOTE: To store async order of the 'response' event
+                    setTimeout(() => {
+                        this._emitOnResponse(cachedResponse.res);
+                    }, 0);
```
Comment (LavrovArtem) [inline src/request-pipeline/destination-request/index.ts:99]:
```suggestion
                    setImmediate(() => this._emitOnResponse(cachedResponse.res));
```

File: src/request-pipeline/incoming-message-like.ts [hunk 5]
```diff
@@ -52,4 +52,26 @@ export default class IncomingMessageLike extends Readable {
...
+    }
+
+    getBody (): Buffer {
+        return this._body;
+    }
+
+    static createFrom (res: IncomingMessage): IncomingMessageLike {
+        const { headers, trailers, statusCode } = res;
+
+        return new IncomingMessageLike({
```
Comment (LavrovArtem) [inline src/request-pipeline/incoming-message-like.ts:67]: also add the `rawHeaders`

**Review by LavrovArtem — APPROVED** [review]

**Pull request merged**